pkg/channel/consolidated/reconciler/controller/consumer_group_watcher.go

```diff
@@ -0,0 +1,197 @@
+	return &WatcherImpl{
+		logger:               logging.FromContext(ctx),
+		adminClient:          ac,
+		pollDuration:         pollDuration,
+		watchers:             make(map[string][]ConsumerGroupHandler),
+		cachedConsumerGroups: make([]string, 0),
+	}
+}
+
+func (w *WatcherImpl) Start() error {
+	w.logger.Infow("ConsumerGroupWatcher starting. Polling for consumer groups", zap.Duration("poll duration", w.pollDuration))
+	go func() {
+		for {
+			select {
+			case <-after(w.pollDuration):
+				// let's get current observed consumer groups
+				observedCGs, err := w.adminClient.ListConsumerGroups()
+				if err != nil {
+					w.logger.Errorw("error while listing consumer groups", zap.Error(err))
+					continue
+				}
+				var notify bool
+				var changedGroup string
+				// Look for observed CGs
+				for _, c := range observedCGs {
+					if !Find(w.cachedConsumerGroups, c) {
+						// This is the first appearance.
+						w.logger.Debugw("Consumer group observed. Caching.",
+							zap.String("consumer group", c))
+						changedGroup = c
+						notify = true
+						break
+					}
+				}
+				// Look for disappeared CGs
+				for _, c := range w.cachedConsumerGroups {
+					if !Find(observedCGs, c) {
+						// This CG was cached but it's no longer there.
+						w.logger.Debugw("Consumer group deleted.",
+							zap.String("consumer group", c))
+						changedGroup = c
+						notify = true
+						break
+					}
+				}
+				if notify {
+					cacheMtx.Lock()
+					w.cachedConsumerGroups = observedCGs
+					cacheMtx.Unlock()
+					w.notify(changedGroup)
+				}
+				done()
+			case <-w.done:
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+func (w *WatcherImpl) Terminate() {
+	w.watchers = nil
+	w.cachedConsumerGroups = nil
+	w.done <- struct{}{}
+}
+
+// TODO explore returning a channel instead of a taking callback
+func (w *WatcherImpl) Watch(watcherID string, cb ConsumerGroupHandler) error {
+	w.logger.Debugw("Adding a new watcher", zap.String("watcherID", watcherID))
+	watchersMtx.Lock()
+	defer watchersMtx.Unlock()
+	cbs, ok := w.watchers[watcherID]
+	if !ok {
+		cbs = []ConsumerGroupHandler{}
+	}
+	w.watchers[watcherID] = append(cbs, cb)
+	// notify at least once to get the current state
+	cb()
+	return nil
+}
+
+func (w *WatcherImpl) Forget(watcherID string) {
+	w.logger.Debugw("Forgetting watcher", zap.String("watcherID", watcherID))
+	delete(w.watchers, watcherID)
+}
+
+func (w *WatcherImpl) List(matcher Matcher) []string {
+	w.logger.Debug("Listing consumer groups")
+	cacheMtx.RLock()
+	defer cacheMtx.RUnlock()
+	cgs := make([]string, 0)
+	for _, cg := range w.cachedConsumerGroups {
+		if matcher(cg) {
+			cgs = append(cgs, cg)
+		}
+	}
+	return cgs
+}
+
+func (w *WatcherImpl) notify(cg string) {
+	watchersMtx.RLock()
+	cacheMtx.RLock()
+	defer watchersMtx.RUnlock()
+	defer cacheMtx.RUnlock()
+
+	for _, cbs := range w.watchers {
+		for _, w := range cbs {
+			w()
+		}
```

Comment:
I think we need to select watchers to notify based on `cg`.

Reply:
as agreed, will use a matcher on a later PR